### DISCUSSION TITLE
Proposals (mainly breaking ones) for 3.0

### DISCUSSION THREAD

**Discussion by rusackas** [apache/superset#19597] [Ideas]
We want to start gathering some _attainable_ wishlist items for 3.0. It's early at this point, so maybe we can gather ideas here (as threads for each proposal), and when we get to the right place in terms of calendar and wishlist, we can open a SIP and a project board. Jump on in!

I'm saying these are _mainly breaking_ changes because things that are _not_ considered breaking can happen at any time and are not held back by a major version release. However, we MAY want to consider some larger/impactful features as a major version milestone, and have a more feature-driven release.

**Comment by rusackas** [comment]
Deprecate drag and drop feature flags: `ENABLE_EXPLORE_DRAG_AND_DROP` & `ENABLE_DND_WITH_CLICK_UX` and remove any code paths possible around the old UX when these flags were disabled.

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by rusackas** [comment]
Deprecate and remove the `UX_BETA` feature flag.

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by rusackas** [comment]
Set `GENERIC_CHART_AXES` Feature flag to `True` by default

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by rusackas** [comment]
Remove Filter Box component

**Reply by rea725** [reply]
@rusackas what replaces the FIlter Box component?  Any writeup on this that you could link to here?

**Reply by rea725** [reply]
Nevermind!  I see the new goodness in master!

**Reply by rea725** [reply]
I am going to miss some aspects of the current filterbox functionality:
1.) I like the new filter pane overall, but for some of our use cases, it would be nice to keep the filters in main dashboard layout.  I'm wondering the there might be a way for the new filters to also have an option be moved about, even if the default behavior is to position them in the filter pane.
2.) In my initial exploring latest master, I didn't see an option to keep the Instant Filtering behavior that exists with the legacy filters.  I really like the instant gratification that comes with being able to make a change and have the dashboard update instantly without the extra step of clicking "Apply Filters" at the bottom.  I think that for small datasets or dashboards that are run on performant databases (such as a Greenplum cluster), the instant filtering option remains very desirable.

**Reply by kasiazjc** [reply]
@rea725 We are actually planning to introduce more flexible filtering (adding filters directly to the dashboard) and we have it planned as a Stage 2 of dashboard filters redesign. We just need to define when it will happen roadmap-wise. 
I think there was a possibility to instantly apply some dashboard filters, but the functionality was confusing before and we need to revisit!

**Reply by kamalkeshavani-aiinside** [reply]
+1 for allowing Instant Filtering in the new setup.
Also, even though filter pane is good, users always like filters next to the charts when that filter is just for that one chart. So the native filter experimental feature should be further explored and allow filters to be inside the dashboard as well.

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1) for consideration/consensus

**Comment by dpgaspar** [comment]
Remove all model views service side rendering code, replace by React list views

**Reply by michael-s-molina** [reply]
Very important for Theming as well.

**Reply by nytai** [reply]
the legacy datasource editor still has some behavior that is missing in the new model based editor. 

eg, https://github.com/apache/superset/discussions/19768

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by dpgaspar** [comment]
Remove all `superset/*` REST APIs marked has deprecated

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by dpgaspar** [comment]
Make all API v1 resource names plural. Ex: `/api/v1/dashboard/` -> `/api/v1/dashboards/`

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by zhaoyongjie** [comment]
Move time range to adhoc filter and move time grain to columns

**Reply by rusackas** [reply]
Can you elaborate more on the "move time grain to columns" aspect?

**Comment by zhaoyongjie** [comment]
Move Advanced Analysis to Metrics

**Reply by rusackas** [reply]
Can you elaborate? Do you have time/intent to contribute this still?

**Comment by cdmikechen** [comment]
I wonder if this feature can be added to the new version as a new extension of superset
[Let Superset provide another canvas based presentation/让Superset提供另一种基于画布方式的展现形式](https://github.com/apache/superset/discussions/18526)

**Reply by rusackas** [reply]
This is a super exciting proposal, but I think it's far too big of a change for the next (imminent) 3.0 release. I'd love to see more traction on the Discussion thread and on Slack, and get more design input to figure out how we can make a change like this gradually. It'd be a HUGE change, so what might be the incremental steps? This is probably best to put out there as a SIP with a full implementation plan if possible.

**Comment by eschutho** [comment]
remove OLD_API_CHECK_DATASET_OWNERSHIP

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by eschutho** [comment]
Create export/import capabilities : roles / users /row level security

**Reply by eschutho** [reply]
@amitmiran137

**Reply by dpgaspar** [reply]
note: FAB can export and import roles and permissions already: https://flask-appbuilder.readthedocs.io/en/latest/cli.html

flask fab export-roles --path roles.json

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by eschutho** [comment]
Remove Pivot Table in favor of Pivot Table V2

**Reply by eschutho** [reply]
@rusackas

**Reply by rusackas** [reply]
Heck yes... I'm not sure of all the breaking changes, but a couple of reported ones include:
• V2 renders NULL/NaN in cells, where V1 displays empty cells
• V2 does not render HTML in cells, where V1 does.

**Comment by rusackas** [comment]
Remove the `/csrf_token/` route from `core.py`, as instructed by a todo comment there

**Reply by rusackas** [reply]
@suddjian for added context

**Reply by rusackas** [reply]
I'm not even sure what this refers to anymore. I don't see the TODO. Am I missing something?

**Comment by eschutho** [comment]
Remove react-select

**Reply by eschutho** [reply]
@michael-s-molina

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

Requires the removal of the FilterBox component.

**Comment by rusackas** [comment]
[Create missing APIs : Roles/Users/ row level security #17487](https://github.com/apache/superset/issues/17487)

**Reply by dpgaspar** [reply]
note: Since version 4.1.0 FAB already has a security API that exposes CRUD operations on Users, roles, permissions. This is disabled by default and still considered beta, to enable set `FAB_ADD_SECURITY_API = True`.

Thank you, once more to @mayurnewase for this contribution ^^^

**Reply by rusackas** [reply]
@dpgaspar is there still a relevant change here to be made for 3.0? If so, what exactly is the current state of affairs? And do we need to keep the above/linked issue open still, or shall we close it?

**Comment by eschutho** [comment]
Remove Legacy Datasource Editor

**Reply by nytai** [reply]
mentioned this above, but
https://github.com/apache/superset/discussions/19768

**Reply by rusackas** [reply]
@eschutho do you think we'll be able to accommodate Tai's request here by the time 3.0 is ready (in Spring 2023?). @nytai is this something you'd still consider a blocker?

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Reply by eschutho** [reply]
Oh, we're working on the new dataset editor now. Let me run this by @yousoph and see if we can squeeze changing a column type in.

**Comment by rusackas** [comment]
[Cleanup] Follow through on removal of Legacy Druid Connector

**Reply by eschutho** [reply]
The breaking change went into 2.0 but the code needs to be cleaned up (part of SIP 68)

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by rusackas** [comment]
[Remove OLD_API_CHECK_DATASET_OWNERSHIP](https://github.com/apache/superset/issues/18907)

**Reply by rusackas** [reply]
Disregarding... I can't seem to find that string in current-era Superset.

**Comment by cemremengu** [comment]
Deprecate `ALERT_REPORTS` feature flag

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Reply by rusackas** [reply]
@cemremengu / @eschutho do you think it makes more sense to just make it default True, or to actually remove the flag entirely?

**Reply by eschutho** [reply]
It's still useful for feature gating on different users, rather than on an application level, so I would vote for keeping it in and defaulting to true if necessary. As a feature it's stable, but it also requires a fair amount of setup for superset admins, so I'm curious if we actually want that feature on by default. I would love to hear from more people perhaps.

**Comment by villebro** [comment]
Remove `viz.py` i.e. migrate the remaining charts to ECharts and/or chart data endpoint. Also remove deprecated fields on `QueryFormData`, `QueryObject` and `QueryContext`, like `groupby`, `metric`, `is_timeseries` etc.

**Reply by rusackas** [reply]
We should break this out into separate items... this is hard to track! But, I'm all for deprecating legacy charts one by one and chipping away at this until we get there!

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

This still needs to be subdivided into individual tasks around which we can gather consensus.

**Comment by RAhuja2015** [comment]
Creating permission role for `Sub-admin`, who can map read/write access to users not beyond `sub-admins` own permissions level.

**Reply by rusackas** [reply]
Can you elaborate more on this? And does it constitute a breaking change in any way?

**Comment by RAhuja2015** [comment]
Feature to add `tags` to dashboard or folders to categorize published dashboards/queries

**Reply by rusackas** [reply]
There's an open PR on this... it's coming soon!

**Reply by rusackas** [reply]
This hasn't merged yet, but is a non-breaking feature.

**Comment by cemremengu** [comment]
Add a path prefix like `/superset`  to all pages for making path handling easier in reverse-proxied environments. i.e. will prevent conflicts with other apps/paths

**Reply by dpgaspar** [reply]
We could probably make this optional so it would not be a breaking change

**Reply by rusackas** [reply]
If it's not a breaking change, I'm not going to add it to the 3.0 project board right now. We can take the PR whenever the time is right.

**Reply by itziarmj** [reply]
Allow customization of the prefix /superset added to all pages, so that several Supersets with different contents could be easily added to the same domain.
This is doable for example in Pentaho Dashboards.

**Comment by cemremengu** [comment]
Replace pandas with [polars](https://github.com/pola-rs/polars)

**Reply by rusackas** [reply]
Could you explain why, or if this constitutes a breaking change?

**Reply by cwegener** [reply]
@rusackas Let me try and explain.
Pandas does not scale well as the number of rows returned from the DB increases, slowing down post-processing in SuperSet.
https://pola-rs.github.io/polars-book/user-guide/#performance-

**EDIT** Maybe worth to get an opinion from @betodealmeida  as well.

**Comment by tiexiw** [comment]
Hope to integrate Alibaba's chart antv, compatible with H5 and mobile phones

**Reply by rusackas** [reply]
We're still working on ECharts completion, which was voted in as the "official" viz library. Superset's mobile capabilities are... well... not quite there yet. But if you're interested in helping to build new plugins, I'd love to speak with you on Slack!

**Comment by markostefanovic1** [comment]
Caching virtual dataset results - if multiple charts on the dashboard are using the same dataset results (but they just group them differently as required for different charts), then send only one query to the database (1 query per dataset), cache the dataset results and allow charts to use those results in chart creation. This should prevent sending a separate query to the database for each chart when the filters are applied.

**Reply by villebro** [reply]
This would require adding database functionality to Superset, as the full result set of the virtual dataset query would need to be persisted and then having the outer query executed against that result set. In addition to needing to add a SQL engine to Superset, this would be infeasible especially for large datasets (imagine moving 1 TB of query results from the analytical database to the Superset cache and then performing a new query on those results)

**Reply by markostefanovic1** [reply]
I'm not really well versed in this subject, but I think I understand what you mean. 

Additional question - would it make more sense if the analytical database's resources were somehow used for caching virtual tables instead of using Superset's resources, OR to use some 3rd party software as a caching layer between the database and Superset?

**Reply by villebro** [reply]
The majority of databases can do this optimization by themselves. Imagine the following virtual dataset:

```sql
select str_col_1, str_col_2, num_col_1, num_col_2 from my_table where col3 = 'foo'
```

Next imagine we have two charts, one grouping by `str_col_1` and summing on `num_col_1`, and the other doing the same on `str_col_2` and `num_col_2` respectively:

First query executed by Superset:
```sql
SELECT str_col_1, SUM(num_col_1) AS my_metric FROM (select str_col_1, str_col_2, num_col_1, num_col_2 from my_table where col_3 = 'foo') virtual_table GROUP BY str_col_1
```

Second query executed by Superset:
```sql
SELECT str_col_2, SUM(num_col_2) AS my_metric FROM (select str_col_1, str_col_2, num_col_1, num_col_2 from my_table where col_3 = 'foo') virtual_table GROUP BY str_col_2
```

In most modern databases, the subquery is just "syntactic sugar" and will be unrolled to the following queries:

First query:
```sql
SELECT str_col_1, SUM(num_col_1) AS my_metric FROM my_table WHERE col_3 = 'foo' GROUP BY str_col_1
```

Second query:
```sql
SELECT str_col_2, SUM(num_col_2) AS my_metric FROM my_table WHERE col_3 = 'foo' GROUP BY str_col_2
```

In other words, the virtual dataset usually doesn't add any execution overhead compared to a regular physical dataset.

**Comment by HeathLee** [comment]
New export/import method and add a copy dataset method to allow change dataset/dashboard when copy. Current export/import charts/dashboard require chart/dashboard name to contain ASCII only (due to `secure_filename` limit)， if the name use non-ascii like a Chinese character, export or import will not work correctly.

**Reply by cccs-rc** [reply]
Hey @HeathLee, just fyi, there's a discussion and draft PR for copying datasets here: https://github.com/apache/superset/discussions/20245
The author mentioned it in Slack as well: https://apache-superset.slack.com/archives/C014LS99C1K/p1654718739800239

**Reply by rusackas** [reply]
This was merged, just to note it here, so we won't worry about it in the 3.0 conversation going forward.

**Comment by dpgaspar** [comment]
convert `ENABLE_BROAD_ACTIVITY_ACCESS` and `MENU_HIDE_USER_INFO` into feature flags

**Reply by rusackas** [reply]
Added to [proposals board](https://github.com/orgs/apache/projects/201/views/1)

**Comment by sfirke** [comment]
Move the general SEARCH box to be on the left of the other boxes?  I find that it is the most-used of the input fields.
![image](https://user-images.githubusercontent.com/7569808/199744705-a62bbbe6-5a7a-4f17-a3df-92969af0e883.png)

**Reply by cemremengu** [reply]
I opened a PR for this sometime ago https://github.com/apache/superset/pull/19721

**Reply by rusackas** [reply]
This was merged, just to note it here, so we won't worry about it in the 3.0 conversation going forward.

**Reply by eschutho** [reply]
side note on this @sfirke, we're planning to convert these crud view tables to the new antd table that has filtering built in, so you may see more improvements here soon.

**Comment by kamalkeshavani-aiinside** [comment]
Allow increasing horizontal grid columns from current fixed 12.

**Reply by rusackas** [reply]
A lot of folks complain about those columns being too narrow as it is, with the card titles getting truncated, and filter icons competing for room. Can you explain any design changes you'd have in mind here, or the motivations for additional columns? This might be part of a bigger effort (i.e. revisiting the dashboard layout feature holistically), so I'd be curious to learn more.

**Reply by kamalkeshavani-aiinside** [reply]
Here is my thought on how to implement this.(I tried to implement it myself, but couldn't figure it out)
https://github.com/apache/superset/issues/16140#issuecomment-900917393
Motivation is very simple, if I want to add 5 'big number' charts of same width in a row, I can't do that without leaving a blank space in the row. Same goes for 7.

**Reply by rusackas** [reply]
I'm not sure... I guess we could in theory go to 24, but 12 seems to be sort of a best practice in similar tools (datadog springs to mind) as additional resolution may cause as much annoyance as utility. Plus, I *know* we'll start seeing more issues about where the card layout crumbles in such small settings with the current design. My *personal* thought here is that we should start a separate Discussion around what a revamped layout editor for Dashboards might look like. I think such a project might be too big for 3.0, as it could get pretty hairy.

**Reply by rusackas** [reply]
If you want to open a PR to make it configurable, that sounds amenable, so long as the default is 12 like you pointed out in your comment. I do worry about changing that setting causing a migration of sorts and breaking a whole bunch of dashboards. This would need pretty comprehensive tests.

**Comment by rusackas** [comment]
This seems like a long shot, since it's pretty intricate, but *maybe* adding mobile/responsive layout support for Dashboards. Just adding it to the convo here since it keeps cropping up on Slack. We proved in a hackathon that it's both _doable_ and... well... quirky.

**Reply by rusackas** [reply]
3.0 needs to come out sooner rather than later, so I'm effectively rescinding this proposal, and we can revisit at any time in the future, so long as it's a non-breaking change.

**Comment by Synarcs** [comment]
One thing I particularly look in superset, is efficient version control, in an multi tenant environment. For example when some role does changes to chart, it tracks local changes, and once he is done the chart can be committed to a superset registry (same like apache nifi and nifi registry). 
I assume in the registry  the chart query context and raw form data can be saved and when importing from registry associated certification can be asked while importing. I know their are other tools like dbt but expect something from core superset as well.

**Reply by rusackas** [reply]
So essentially version control for charts? This has been discussed in various approaches (also for dashboards and queries), but I think the design was never quite finalized, and nobody ever stepped up to volunteer to actually build such a feature as part of their roadmap. If you have interest in contributing this engineering effort, I'd love to see (and be tagged on) a GIthub Discussion (in the "Ideas" category), so we can take it up as a community.

**Comment by eschutho** [comment]
We can probably remove this from the frontend extension registry:
```
// Exporting this under the old name for backwards compatibility.
// After downstream folks have migrated to `getExtensionsRegistry`, we should remove this.
export const getUiOverrideRegistry = getExtensionsRegistry;
```

**Reply by rusackas** [reply]
Added to the board :)

**Comment by eschutho** [comment]
This has been deprecated, so we can remove it from the docker config: `ENABLE_SCHEDULED_EMAIL_REPORTS`

**Reply by rusackas** [reply]
Added to [the board](https://github.com/orgs/apache/projects/201/views/1).

**Comment by EugeneTorap** [comment]
Move SQL Lab to SPA 🙏 🙏 🙏

**Reply by rusackas** [reply]
Added to [the board](https://github.com/orgs/apache/projects/201/views/1). Are you willing/able to work on this, or is it more of a request?

**Comment by rusackas** [comment]
All items have been moved to the board, which is in flight. We'll take this up in a new thread when 4.0 comes.
